app/src/org/commcare/android/resource/AndroidResourceManager.java

```diff
@@ -0,0 +1,267 @@
+
+import android.content.Context;
+import android.os.Handler;
+import android.util.Log;
+
+import org.commcare.android.javarosa.AndroidLogger;
+import org.commcare.android.resource.analytics.UpdateStatPersistence;
+import org.commcare.android.resource.analytics.UpdateStats;
+import org.commcare.android.resource.installers.LocalStorageUnavailableException;
+import org.commcare.android.tasks.UpdateTask;
+import org.commcare.android.util.AndroidCommCarePlatform;
+import org.commcare.android.util.AndroidResourceInstallerFactory;
+import org.commcare.dalvik.application.CommCareApp;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.commcare.resources.ResourceManager;
+import org.commcare.resources.model.InstallCancelled;
+import org.commcare.resources.model.InstallCancelledException;
+import org.commcare.resources.model.Resource;
+import org.commcare.resources.model.ResourceTable;
+import org.commcare.resources.model.TableStateListener;
+import org.commcare.resources.model.UnresolvedResourceException;
+import org.commcare.util.CommCarePlatform;
+import org.javarosa.core.services.Logger;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+
+/**
+ * Manages app installations and updates. Extends the ResourceManager with the
+ * ability to stage but not apply updates.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class AndroidResourceManager extends ResourceManager {
+    private final static String TAG = AndroidResourceManager.class.getSimpleName();
+    public final static String TEMP_UPGRADE_TABLE_KEY = "TEMP_UPGRADE_RESOURCE_TABLE";
+    private final UpdateStats updateStats;
+    private final CommCareApp app;
+    private String profileRef;
+    private final ResourceTable tempUpgradeTable;
+
+    // 5 minutes
+    private final static long UPDATE_RETRY_DELAY_IN_MS = 1000 * 60 * 5;
+
+    public AndroidResourceManager(AndroidCommCarePlatform platform) {
+        super(platform, platform.getGlobalResourceTable(),
+                platform.getUpgradeResourceTable(),
+                platform.getRecoveryTable());
+
+        app = CommCareApplication._().getCurrentApp();
+
+        tempUpgradeTable =
+                ResourceTable.RetrieveTable(app.getStorage(TEMP_UPGRADE_TABLE_KEY, Resource.class),
+                        new AndroidResourceInstallerFactory(app));
+
+        updateStats = UpdateStatPersistence.loadUpdateStats(app);
+        upgradeTable.setInstallStatsLogger(updateStats);
+        tempUpgradeTable.setInstallStatsLogger(updateStats);
+    }
+
+    /**
+     * Download the latest profile; if it is new, download and stage the entire
+     * update.
+     *
+     * @param profileRef Reference that resolves to the profile file used to
+     *                   seed the update
+     * @return UpdateStaged upon update download, UpToDate if no new update,
+     * otherwise an error status.
+     */
+    public AppInstallStatus checkAndPrepareUpgradeResources(String profileRef) {
```

Comment:
Could make this section a little clearer if this method only did the "check" part -- would need to add another value to AppInstallStatus called like "UpdateNeeded" that would be returned at the end of the try block instead of calling prepareUpgradeResources(). And then in stageUpdate(), you would check for the return value of the call to checkForUpgrade(), and call prepareUpgradeResources() iff it returned AppInstallStatus.UpdateNeeded.